apps/web/src/hooks/useVariablesManager.ts

```diff
@@ -1,64 +1,64 @@
   const variablesArray = useFieldArray({ control, name: `steps.${index}.template.variables` });
   const variableArray = watch(`steps.${index}.template.variables`, []);
 
-  useEffect(() => {
+  const getTextContent = useCallback((templateToParse?: ITemplates): string => {
+    return contents
+      .map((con) => con.split('.').reduce((a, b) => a && a[b], templateToParse ?? {}))
+      .map((con) => (Array.isArray(con) ? con.map((innerCon) => innerCon.content).join(' ') : con))
+      .join(' ');
+  }, []);
+
+  const [textContent, setTextContent] = useState<string>(() => getTextContent(getValues(`steps.${index}.template`)));
+
+  useLayoutEffect(() => {
     const subscription = watch((values) => {
-      gatherTextContent(values.steps[index].template);
+      const steps = values.steps ?? [];
+      if (!steps.length || !steps[index]) return;
+
+      const step = steps[index];
+      setTextContent(getTextContent(step?.template as ITemplates));
     });
 
     return () => subscription.unsubscribe();
-  }, [watch, contents]);
-
-  useEffect(() => {
-    const template = getValues(`steps.${index}.template`);
-    gatherTextContent(template);
-  }, [contents]);
+  }, [index, watch, setTextContent, getTextContent, contents]);
 
-  useMemo(() => {
+  useLayoutEffect(() => {
     try {
-      setAst(parse(textContent));
-    } catch (e) {
-      return;
-    }
-  }, [textContent]);
+      const ast = parse(textContent);
+      const variables = getTemplateVariables(ast.body);
+      const arrayFields = [...(variableArray || [])];
 
-  function gatherTextContent(template = {}) {
-    setTextContent(
-      contents
-        .map((con) => con.split('.').reduce((a, b) => a && a[b], template))
-        .map((con) => (Array.isArray(con) ? con.map((innerCon) => innerCon.content).join(' ') : con))
-        .join(' ')
-    );
-  }
+      variables.forEach((vari) => {
+        if (!arrayFields.find((field) => field.name === vari.name)) {
+          arrayFields.push(vari);
+        }
+      });
 
-  useMemo(() => {
-    const variables = getTemplateVariables(ast.body);
-    const arrayFields = [...(variableArray || [])];
+      arrayFields.forEach((vari, ind) => {
+        if (!variables.find((field) => field.name === vari.name)) {
+          delete arrayFields[ind];
+        }
+      });
 
-    variables.forEach((vari) => {
-      if (!arrayFields.find((field) => field.name === vari.name)) {
-        arrayFields.push(vari);
-      }
-    });
+      const newVariablesArray = arrayFields.filter((field) => !!field);
 
-    arrayFields.forEach((vari, ind) => {
-      if (!variables.find((field) => field.name === vari.name)) {
-        delete arrayFields[ind];
+      if (!isEqual(variableArray, newVariablesArray)) {
+        variablesArray.replace(newVariablesArray);
       }
-    });
-
-    variablesArray.replace(arrayFields.filter((field) => !!field));
-  }, [ast]);
+    } catch (e) {
+      return;
+    }
+  }, [textContent, variableArray]);
```

Comment:
parse the `textContent` and build the new variables array, then replace it in the form if there are differences with the current variables array